rust/private/rustc.bzl

```diff
@@ -1075,32 +1082,219 @@ def construct_arguments(
 
     return args, env
 
+def get_edition(attr, toolchain, label):
+    """Returns the Rust edition from either the current rule's attirbutes or the current `rust_toolchain`
+
+    Args:
+        attr (struct): The current rule's attributes
+        toolchain (rust_toolchain): The `rust_toolchain` for the current target
+        label (Label): The label of the target being built
+
+    Returns:
+        str: The target Rust edition
+    """
+    if getattr(attr, "edition"):
+        return attr.edition
+    elif not toolchain.default_edition:
+        fail("Attribute `edition` is required for {}.".format(label))
```

Comment:
Preexisting, but the first time I hit this error it wasn't obvious that you could set a default edition in the toolchain, might be worth a mention in this error?

Reply:
I'm leaving your comments unresolved in this PR and will address in a separate PR because they're not within scope.